cf/query.py

```diff
@@ -857,32 +857,47 @@ def set_condition_units(self, units):
         <CF Query: (ge 3000 m)>
 
         """
+        def get_and_set_value_units(v, u):
+            """Helper method to simplify logic to set specified units."""
+            v_units = getattr(v, "Units", None)
+            if v_units is None:  # Value 'v' has no units
+                v = Data(v, units=u)
+            else:  # Value 'v' already has units
+                try:
+                    v = v.copy()
+                    v.Units = u
+                except ValueError:
+                    raise ValueError(
+                        f"Units {u!r} are not equivalent to "
+                        f"query condition units {v_units!r}"
+                    )
+
+            return v
+
         units = Units(units)
 
         compound = self._compound
         if compound:
             for r in compound:
                 r.set_condition_units(units)
-
             return
 
         value = self._value
         if value is None:
             return
 
         value_units = getattr(value, "Units", None)
-        if value_units is None:
-            # Value has no units
-            value = Data(value, units=units)
+        if self.operator in ("wi", "wo", "set"):
+            # Value is a sequence of things that may or may not
+            # already have units
```

Comment:
There's a potential performance issue here, in that if a large set is stored as a `Data` object, as opposed to a python built-in like a list, then iterating through it could be slow. For the `Data` object case we could just change the units without iterating through.

My inclination is to resist the overwhelming urge for premature optimsation and do nothing for now, as I confident that folk don't/won't be using huge sets, but I'll leave that decision to you :)